proc/groupby.go

```diff
@@ -303,67 +342,71 @@ func (g *GroupByAggregator) recordsForTable(table map[string]*GroupByRow) []*zso
 	}
 	sort.Strings(keys)
 	var recs []*zson.Record
-Keys:
-	for _, k := range keys {
-		row := table[k]
-		var vals [][]byte
-		if g.TimeBinDuration > 0 {
-			vals = append(vals, []byte(row.ts.StringFloat()))
-		}
-		for _, v := range row.keyVals {
-			vals = append(vals, v)
-		}
-		for _, red := range row.columns.Reducers {
-			v, err := zson.ZvalFromZeekString(nil, reducer.Result(red).String())
-			if err != nil {
-				g.logger.Error("zson.ZvalFromZeekString failed", zap.Error(err))
-				continue Keys
+
+	for _, typeCol := range g.typeCols {
+		for _, k := range keys {
+			row := table[k]
+			if !typeMatch(typeCol, row.rowKeys) {
+				continue
 			}
-			vals = append(vals, v)
-		}
-		d := g.lookupDescriptor(&row.columns)
-		r, err := zson.NewRecordZvals(d, vals...)
-		if err != nil {
-			g.logger.Error("zson.NewRecordZvals failed", zap.Error(err))
+			var zv zval.Encoding
+			if g.TimeBinDuration > 0 {
+				panic("BAR")
+				zv = zval.AppendValue(zv, []byte(row.ts.StringFloat()))
+			}
+			for _, rowkey := range row.rowKeys {
+				zv = zval.Append(zv, rowkey.Body, zeek.IsContainerType(rowkey.Type))
+			}
+			for _, red := range row.columns.Reducers {
+				// a reducer value is never a container
+				v := reducer.Result(red)
+				if zeek.IsContainer(v) {
+					panic("internal bug: reducer result cannot be a container!")
+				}
+				zv = v.Encode(zv)
+			}
+			d := g.lookupDescriptor(row)
+			r := zson.NewRecord(d, row.ts, zv)
+			recs = append(recs, r)
 		}
-		recs = append(recs, r)
 	}
 	return recs
 }
 
-func (g *GroupByAggregator) lookupDescriptor(columns *compile.Row) *zson.Descriptor {
-	keyCols := make([]zeek.Column, len(columns.Reducers))
-	for k, red := range columns.Reducers {
-		z := reducer.Result(red)
-		keyCols[k] = zeek.Column{
-			Name: columns.Defs[k].Target(),
-			Type: z.Type(),
-		}
-	}
-	outcols := append(g.staticCols, keyCols...)
-	return g.dt.GetByColumns(outcols)
-}
-
 // initialize the static columns, namely the td, ts (if time-binned), and key columns.
-func (g *GroupByAggregator) initStaticCols(r *zson.Record) {
+func (g *GroupByAggregator) appendColumns(columns []zeek.Column, rowkeys []zeek.TypedEncoding, defs compile.Row) []zeek.Column {
```

Comment:
This appears to be unused?

Reply:
Yes, that was the old bad stuff.  Deleted.